hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/proxy/SCMBlockLocationFailoverProxyProvider.java

```diff
@@ -17,283 +17,34 @@
-import org.apache.hadoop.io.retry.RetryPolicies;
-import org.apache.hadoop.io.retry.RetryPolicy;
-import org.apache.hadoop.ipc.ProtobufRpcEngine;
-import org.apache.hadoop.ipc.RPC;
-import org.apache.hadoop.net.NetUtils;
-import org.apache.hadoop.security.UserGroupInformation;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Failover proxy provider for SCM block location.
  */
-public class SCMBlockLocationFailoverProxyProvider implements
-    FailoverProxyProvider<ScmBlockLocationProtocolPB>, Closeable {
+public class SCMBlockLocationFailoverProxyProvider extends
+    SCMFailoverProxyProviderBase<ScmBlockLocationProtocolPB> {
   public static final Logger LOG =
       LoggerFactory.getLogger(SCMBlockLocationFailoverProxyProvider.class);
-  private final SCMClientConfig scmClientConfig;
-
-  private final Map<String, ProxyInfo<ScmBlockLocationProtocolPB>> scmProxies;
-  private final Map<String, SCMProxyInfo> scmProxyInfoMap;
-  private List<String> scmNodeIds;
-
-  // As SCM Client is shared across threads, performFailOver()
-  // updates the currentProxySCMNodeId based on the updateLeaderNodeId which is
-  // updated in shouldRetry(). When 2 or more threads run in parallel, the
-  // RetryInvocationHandler will check the expectedFailOverCount
-  // and not execute performFailOver() for one of them. So the other thread(s)
-  // shall not call performFailOver(), it will call getProxy() which uses
-  // currentProxySCMNodeId and returns the proxy.
-  private volatile String currentProxySCMNodeId;
-  private volatile int currentProxyIndex;
-
-  private final ConfigurationSource conf;
-  private final long scmVersion;
-
-  private String scmServiceId;
-
-  private final int maxRetryCount;
-  private final long retryInterval;
-
-  private final UserGroupInformation ugi;
-
-  private String updatedLeaderNodeID = null;
-
 
   public SCMBlockLocationFailoverProxyProvider(ConfigurationSource conf) {
-    this.conf = conf;
-    this.scmVersion = RPC.getProtocolVersion(ScmBlockLocationProtocolPB.class);
-
-    try {
-      this.ugi = UserGroupInformation.getCurrentUser();
-    } catch (IOException ex) {
-      LOG.error("Unable to fetch user credentials from UGI", ex);
-      throw new RuntimeException(ex);
-    }
-
-    // Set some constant for non-HA.
-    if (scmServiceId == null) {
-      scmServiceId = SCM_DUMMY_SERVICE_ID;
-    }
-    this.scmProxies = new HashMap<>();
-    this.scmProxyInfoMap = new HashMap<>();
-
-    loadConfigs();
-
-    this.currentProxyIndex = 0;
-    currentProxySCMNodeId = scmNodeIds.get(currentProxyIndex);
-
-    scmClientConfig = conf.getObject(SCMClientConfig.class);
-    this.maxRetryCount = scmClientConfig.getRetryCount();
-    this.retryInterval = scmClientConfig.getRetryInterval();
-
+    super(ScmBlockLocationProtocolPB.class, conf, null);
     LOG.info("Created block location fail-over proxy with {} nodes: {}",
-        scmNodeIds.size(), scmProxyInfoMap.values());
```

Comment:
I think this log message is no longer needed, since base class also logs the same.

Reply:
Well spotted. Updated.